File: wiz-event-mall-java/src/main/java/com/example/wizeventmalljava/api/event/EventService.java
package com.example.wizeventmalljava.api.event;

import java.time.Duration;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.example.wizeventmalljava.api.event.dto.EventCreateRequest;
import com.example.wizeventmalljava.api.event.dto.EventInfoResponse;
import com.example.wizeventmalljava.api.event.dto.EventParticipateResponse;
import com.example.wizeventmalljava.domain.event.entity.Event;
import com.example.wizeventmalljava.domain.event.entity.EventWinner;
import com.example.wizeventmalljava.domain.event.repository.EventRepository;
import com.example.wizeventmalljava.domain.event.repository.EventWinnerRepository;
import com.example.wizeventmalljava.domain.lock.RedissonLockService;
import com.example.wizeventmalljava.domain.member.repository.MemberRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class EventService {

	private final EventRepository eventRepository;
	private final EventWinnerRepository eventWinnerRepository;
	private final MemberRepository memberRepository;
	private final RedissonLockService redissonLockService;

	public Long create(EventCreateRequest req) {
		Event event = eventRepository.save(req.toEntity());
		return event.getId();
	}

	public EventInfoResponse retrieve(Long eventId) {
		return eventRepository.findById(eventId)
			.map(EventInfoResponse::from)
			.orElseThrow(() -> new RuntimeException("존재하지 않는 이벤트입니다. (" + eventId + ")"));
	}

	@Transactional
	public EventParticipateResponse participate(Long eventId, Long memberId) {
		if (!memberRepository.existsById(memberId)) {
			throw new IllegalArgumentException("잘못된 사용자 정보입니다.");
		}
		if (eventWinnerRepository.existsByEventIdAndMemberId(eventId, memberId)) {
			throw new IllegalStateException("이미 이벤트에 참여하셨습니다.");
		}

		Event event = eventRepository.findById(eventId)
			.orElseThrow(() -> new RuntimeException("이벤트를 찾을 수 없습니다. (" + eventId + ")"));
		long winnerCount = eventWinnerRepository.countByEvent(event);
		if (!event.possibleParticipate(winnerCount)) {
			throw new IllegalStateException("이벤트 참여가 불가능합니다.");
		}

		EventWinner eventWinner = eventWinnerRepository.save(event.win(memberId));
		return EventParticipateResponse.from(eventWinner);
	}

	@Transactional
	public EventParticipateResponse participateWithLock(Long eventId, Long memberId) {
		return redissonLockService.lock(
			"PARTICIPATE:" + eventId,
			Duration.ofSeconds(60),
			() -> participate(eventId, memberId));
	}
}

File: wiz-event-mall/src/main/kotlin/com/example/wizeventmall/api/event/EventService.kt
package com.example.wizeventmall.api.event

import com.example.wizeventmall.api.event.dto.EventInfoResponse
import com.example.wizeventmall.api.event.dto.EventParticipateResponse
import com.example.wizeventmall.api.event.dto.EventRegisterRequest
import com.example.wizeventmall.domain.event.repository.EventRepository
import com.example.wizeventmall.domain.event.repository.EventWinnerRepository
import com.example.wizeventmall.domain.lock.RedissonLock
import com.example.wizeventmall.domain.member.repository.MemberRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Isolation
import org.springframework.transaction.annotation.Transactional

@Service
class EventService(
    private val eventRepository: EventRepository,
    private val eventWinnerRepository: EventWinnerRepository,
    private val memberRepository: MemberRepository
) {

    fun register(req: EventRegisterRequest): Long {
        return eventRepository.save(req.toEntity())
            .let { it.id!! }
    }

    fun retrieve(eventId: Long): EventInfoResponse {
        return eventRepository.findByIdOrNull(eventId)
            ?.let { EventInfoResponse.from(it) }
            ?: throw RuntimeException("존재하지 않는 이벤트입니다. ($eventId)")
    }

//    @Transactional
//    fun participate(eventId: Long, memberId: Long): EventParticipateResponse {
//        check(memberRepository.existsById(memberId)) { "잘못된 사용자 정보입니다." }
//        check(!eventWinnerRepository.existsByEventIdAndMemberId(eventId, memberId)) { "이미 이벤트에 참여하셨습니다." }
//
//        return eventRepository.findByIdOrNull(eventId)!!.takeIf {
//            val winnerCount = eventWinnerRepository.countByEvent(it)
//            it.possibleParticipate(currentWinnerCount = winnerCount)
//        }?.let {
//            eventWinnerRepository.save(it.win(memberId))
//        }.let {
//            EventParticipateResponse.from(it)
//        }
//    }

    @Transactional
    fun participate(eventId: Long, memberId: Long): EventParticipateResponse = RedissonLock("PARTICIPATE:$eventId") {
        check(memberRepository.existsById(memberId)) { "잘못된 사용자 정보입니다." }
        check(!eventWinnerRepository.existsByEventIdAndMemberId(eventId, memberId)) { "이미 이벤트에 참여하셨습니다." }

        eventRepository.findByIdOrNull(eventId)!!.takeIf {
            val winnerCount = eventWinnerRepository.countByEvent(it)
            it.possibleParticipate(currentWinnerCount = winnerCount)
        }?.let {
            eventWinnerRepository.save(it.win(memberId))
        }.let {
            EventParticipateResponse.from(it)
        }
    }
}